### DISCUSSION TITLE
output format (type) of gatherer 'corosync.conf' when requesting e.g. 'totem.interface'

### DISCUSSION THREAD

**Discussion by angelabriel** [trento-project/wanda#125] [Q&A] [answered]
I have seen some inconsistency (?) in the response of
`/usr/bin/trento-agent facts gather --gatherer "corosync.conf" --argument "totem.interface"`

When /etc/corosync/corosync.conf contains **two** interfaces, the gatherer returns an array holding 2 objects like
`"value":[{"mcastport":5405.0,"ringnumber":0.0,"ttl":1.0},{"mcastport":5407.0,"ringnumber":1.0,"ttl":1.0}]`

If only **one** interface is defined, the gatherer returns an object and not an array holding only one object (my expectation)
`"value":{"mcastport":5405.0,"ringnumber":0.0,"ttl":1.0}`

And defining **three** interfaces, the gatherer returns an array holding **two** elements, the **first** element is an array holding 2 objects and the **second** element is an object.
`"value":[[{"mcastport":5405.0,"ringnumber":0.0,"ttl":1.0},{"mcastport":5407.0,"ringnumber":1.0,"ttl":1.0}],{"mcastport":5409.0,"ringnumber":2.0,"ttl":1.0}]`

My expectation was to get an array holding 3 objects

When trying to use some more complex processing for the expectations like
```
    expect: |
        let rings = 0;
        for entry in facts.corosync_interfaces {
            if entry.ringnumber != "" {
                rings = rings + 1;
            }
        }
        if rings == values.expected_corosync_rings_per_node {
            true
        }
```

How can this be done, if the resulting datatype of 'facts.corosync_interfaces' is changing regarding the content of the retrieved config file?

So is this behavior intentional? What do I missing?

The same happens for` --argument "nodelist.node"` and even that we today only need 2 nodes for the supported scenarios for the future (with support for SAPHana 'ScaleOut') we need to support more than 2 entries in a predictable way.

Additional info:
`# /usr/bin/trento-agent facts gather --gatherer "corosync.conf" --argument "totem.interface.0.ringnumber"`
and
`# /usr/bin/trento-agent facts gather --gatherer "corosync.conf" --argument "totem.interface.1.ringnumber"`
are working well, if the config file does contain 2 interface entries.
When containing 3 interfaces, only
`# /usr/bin/trento-agent facts gather --gatherer "corosync.conf" --argument "totem.interface.1.ringnumber"`
is working, but
`# /usr/bin/trento-agent facts gather --gatherer "corosync.conf" --argument "totem.interface.0.ringnumber"`
is returning an error

**Comment by angelabriel** [comment]
And a second (related) question:
Why are the values of the entries contain a trailing '**.0**'?
`{"mcastport":5405.0,"ringnumber":0.0,"ttl":1.0}`
But in the config file it's
```
interface {
                ringnumber: 0
                mcastport: 5405
                ttl: 1
        }
```
Intentional?
If yes, we should extend the documentation to mention this, because we need to set the 'expected' values in the checks accordingly. 

Note: It's not the gatherer, who added the '**.0**'  as a 

```
# /usr/bin/trento-agent facts gather --gatherer "corosync.conf" --argument "totem.interface"
INFO[2022-12-12 18:48:57] Using config file: /etc/trento/agent.yaml    
INFO[2022-12-12 18:48:57] loading plugins                              
INFO[2022-12-12 18:48:57] Starting corosync.conf file facts gathering process 
INFO[2022-12-12 18:48:57] Requested corosync.conf file facts gathered  
INFO[2022-12-12 18:48:57] Gathered fact for "corosync.conf" with argument "totem.interface": 
INFO[2022-12-12 18:48:57] {
  "Name": "totem.interface",
  "CheckID": "",
  "Value": {
    "Value": [
      {
        "Value": {
          "mcastport": {
            "Value": 5405
          },
          "ringnumber": {
            "Value": 0
          },
          "ttl": {
            "Value": 1
          }
        }
      },
      {
        "Value": {
          "mcastport": {
            "Value": 5407
          },
          "ringnumber": {
            "Value": 1
          },
          "ttl": {
            "Value": 1
          }
        }
      }
    ]
  },
  "Error": null
} 
```
shows. It's added during the evaluation of the gatherer result inside wanda?

**Reply by fabriziosestito** [reply]
~this is not intentional and I am tracking it as a bug~

**Reply by fabriziosestito** [reply]
@angelabriel on second thought, this shouldn't matter since for our evaluation language `1 == 1.0`

**Reply by angelabriel** [reply]
Thank you @fabriziosestito This is good to know. I probably overlooked it in the rhai documentation.

**Comment by fabriziosestito** [comment]
most probably a bug, I am tracking this and we'll get you ASAP

**Comment by arbulu89** [comment, marked as answer]
@angelabriel Fix here: https://github.com/trento-project/agent/pull/155

**Reply by angelabriel** [reply]
yeah, just saw it. Looks pretty good. Many thanks

**Reply by angelabriel** [reply]
tested the fixed gatherer with new trento-agent package trento-agent-1.2.0+git.dev24.1671109436.b619a69-150300.1.1.x86_64.rpm 
Works like a charm. Many thanks for the fast fix.
